File: backends/java/zappy-domain/src/main/java/com/zappymart/domain/accounts/PasswordPolicy.java
package com.zappymart.domain.accounts;

import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserErrorCode;

public final class PasswordPolicy {

    public static final int SHORTEST = 12;

    public static final int LONGEST = 128;

    private PasswordPolicy() {
    }

    public static Result<String> check(String password) {
        if (password == null || password.length() < SHORTEST) {
            return Result.refuse(UserErrorCode.PASSWORD_TOO_SHORT,
                    "A password is at least " + SHORTEST + " characters long.", "input.password");
        }
        if (password.length() > LONGEST) {
            return Result.refuse(UserErrorCode.PASSWORD_TOO_LONG,
                    "A password is at most " + LONGEST + " characters long.", "input.password");
        }
        return Result.of(password);
    }
}

File: backends/kotlin/zappy-domain/src/main/kotlin/nl/zappymart/domain/accounts/PasswordPolicy.kt
package nl.zappymart.domain.accounts

import nl.zappymart.domain.shared.Result
import nl.zappymart.domain.shared.UserErrorCode
import nl.zappymart.domain.shared.refusal

object PasswordPolicy {

    const val MINIMUM_LENGTH = 12

    const val MAXIMUM_LENGTH = 128

    fun check(password: String): Result<Unit> = when {
        password.length < MINIMUM_LENGTH -> refusal(
            UserErrorCode.PASSWORD_TOO_SHORT,
            "A password is at least $MINIMUM_LENGTH characters long.",
            "input.password",
        )

        password.length > MAXIMUM_LENGTH -> refusal(
            UserErrorCode.PASSWORD_TOO_LONG,
            "A password is at most $MAXIMUM_LENGTH characters long.",
            "input.password",
        )

        else -> Result.Success(Unit)
    }
}
